File: app/src/main/java/com/team2052/frckrawler/metric/IndexValueMetricType.java
package com.team2052.frckrawler.metric;

import com.google.common.collect.Maps;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.team2052.frckrawler.data.tba.JSON;
import com.team2052.frckrawler.helpers.Tuple2;
import com.team2052.frckrawler.helpers.metric.MetricDataHelper;
import com.team2052.frckrawler.metric.data.CompiledMetricValue;
import com.team2052.frckrawler.metric.data.MetricValue;
import com.team2052.frckrawler.metric.view.ListIndexMetricWidget;
import com.team2052.frckrawler.models.Metric;

import java.util.List;
import java.util.Map;

public abstract class IndexValueMetricType<M extends ListIndexMetricWidget> extends MetricType<M> {
    private static JsonObject compiledValueToJson(Map<Integer, Tuple2<String, Double>> compiledMap, JsonArray possible_values) {
        JsonObject compiledValue = new JsonObject();
        JsonArray values = JSON.getGson().toJsonTree(Tuple2.yieldValues(compiledMap.values()).toArray()).getAsJsonArray();
        compiledValue.add("names", possible_values);
        compiledValue.add("values", values);
        return compiledValue;
    }

    @Override
    public CompiledMetricValue compile(Metric metric, List<MetricValue> metricValues, float weight) {
        final JsonArray possibleValues = JSON.getAsJsonObject(metric.getData()).get("values").getAsJsonArray();
        final Map<Integer, Tuple2<String, Double>> compiledMap = Maps.newTreeMap();
        double denominator = 0.0;

        for (int i = 0; i < possibleValues.size(); i++) {
            compiledMap.put(i, new Tuple2<>(possibleValues.get(i).getAsString(), 0.0));
        }

        if (metricValues.isEmpty()) {
            return CompiledMetricValue.create(metric, metricValues, compiledValueToJson(compiledMap, possibleValues));
        }

        for (int i = 0; i < metricValues.size(); i++) {
            MetricValue value = metricValues.get(i);
            Tuple2<List<Integer>, MetricDataHelper.ReturnResult> value_result = MetricDataHelper.getListIndexMetricValue(value);
            if (value_result.t2.isError) {
                continue;
            }

            double weightForMatch = MetricDataHelper.getCompileWeightForMatchNumber(value, metricValues, weight);

            for (Integer index : value_result.t1) {
                if (!compiledMap.containsKey(index)) {
                    continue;
                }

                double newValue = compiledMap.get(index).t2 + weightForMatch;
                compiledMap.put(index, compiledMap.get(index).setT2(newValue));
            }

            denominator += weightForMatch;
        }

        for (Map.Entry<Integer, Tuple2<String, Double>> entry : compiledMap.entrySet()) {
            double percent = Math.round((entry.getValue().t2 / denominator * 100) * 100.0)
                    / 100.0;
            compiledMap.put(entry.getKey(), entry.getValue().setT2(percent));
        }

        return CompiledMetricValue.create(metric, metricValues, compiledValueToJson(compiledMap, possibleValues));
    }

    @Override
    public String convertCompiledValueToString(JsonObject jsonObject) {
        JsonArray names = jsonObject.get("names").getAsJsonArray();
        JsonArray values = jsonObject.get("values").getAsJsonArray();
        StringBuilder value = new StringBuilder();
        for (int i = 0; i < names.size(); i++) {
            value.append(String.format("%s:%s%s\n", names.get(i).getAsString(), values.get(i).getAsDouble(), '%'));
        }
        return value.toString();
    }
}

File: app/src/main/java/com/team2052/frckrawler/metric/IndexValueMetricType.kt
package com.team2052.frckrawler.metric

import com.google.common.collect.Maps
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.team2052.frckrawler.data.tba.v3.JSON
import com.team2052.frckrawler.helpers.Tuple2
import com.team2052.frckrawler.helpers.metric.MetricDataHelper
import com.team2052.frckrawler.metric.data.CompiledMetricValue
import com.team2052.frckrawler.metric.data.MetricValue
import com.team2052.frckrawler.metric.view.ListIndexMetricWidget
import com.team2052.frckrawler.models.Metric

abstract class IndexValueMetricType<M : ListIndexMetricWidget> : MetricType<M>() {
    private fun compiledValueToJson(compiledMap: Map<Int, Tuple2<String, Double>>, possible_values: JsonArray): JsonObject {
        val compiledValue = JsonObject()
        val values = JSON.getGson().toJsonTree(Tuple2.yieldValues(compiledMap.values).toTypedArray()).asJsonArray
        compiledValue.add("names", possible_values)
        compiledValue.add("values", values)
        return compiledValue
    }

    override fun compile(metric: Metric, metricValues: List<MetricValue>, weight: Float): CompiledMetricValue {
        val possibleValues = JSON.getAsJsonObject(metric.data).get("values").asJsonArray
        val compiledMap = Maps.newTreeMap<Int, Tuple2<String, Double>>()
        var denominator = 0.0

        for (i in 0..possibleValues.size() - 1) {
            compiledMap.put(i, Tuple2(possibleValues.get(i).asString, 0.0))
        }

        if (metricValues.isEmpty()) {
            return CompiledMetricValue(
                    metric = metric,
                    metricValues = metricValues,
                    jsonValue = compiledValueToJson(compiledMap, possibleValues)
            )
        }

        for (i in metricValues.indices) {
            val value = metricValues[i]
            val value_result = MetricDataHelper.getListIndexMetricValue(value)
            if (value_result.t2.isError) {
                continue
            }

            val weightForMatch = MetricDataHelper.getCompileWeightForMatchNumber(value, metricValues, weight.toDouble())

            for (index in value_result.t1) {
                if (!compiledMap.containsKey(index)) {
                    continue
                }

                val newValue = compiledMap[index]?.t2?.plus(weightForMatch)
                compiledMap.put(index, compiledMap[index]?.setT2(newValue))
            }

            denominator += weightForMatch
        }

        for ((key, value) in compiledMap) {
            val percent = Math.round(value.t2 / denominator * 100 * 100.0) / 100.0
            compiledMap.put(key, value.setT2(percent))
        }

        return CompiledMetricValue(
                metric = metric,
                metricValues = metricValues,
                jsonValue = compiledValueToJson(compiledMap, possibleValues))
    }

    override fun convertCompiledValueToString(jsonObject: JsonObject): String {
        val names = jsonObject.get("names").asJsonArray
        val values = jsonObject.get("values").asJsonArray
        val value = StringBuilder()
        for (i in 0..names.size() - 1) {
            value.append(String.format("%s:%s%s\n", names.get(i).asString, values.get(i).asDouble, '%'))
        }
        return value.toString()
    }
}
